### PR TITLE
[Reviewer: Rob] Fixes iss1267: Adds a tel URI alias to the P-Asserted-Identity header

### PR DESCRIPTION
**Pull request by AME2**
Fixes iss1267 
https://github.com/Metaswitch/sprout/issues/1267
Instead of only adding a tel URI alias if it matches the user of the SIP URI, we now check if there is a tel URI alias whose username matches the SIP URI. If there is we add it to the P-Asserted-Identity header and if there isn't we add the first tel URI in the alias list.

### CODE REVIEW

File: src/scscfsproutlet.cpp [hunk 1]
```diff
@@ -1825,6 +1828,23 @@ void SCSCFSproutletTsx::add_second_p_a_i_hdr(pjsip_msg* msg)
                                        new_p_a_i_str,
                                        asserted_id->name_addr.display);
       }
+      else
+      {
+        for (unsigned int i=0; i<_aliases.size(); i++)
```
Comment (rkd-msw) [inline src/scscfsproutlet.cpp]: We'd normally iterate over a vector using a C++ iterator - see https://github.com/Metaswitch/sprout/blob/dev/src/aschain.cpp#L267 for an example.

File: src/scscfsproutlet.cpp [hunk 2]
```diff
@@ -1825,6 +1828,23 @@ void SCSCFSproutletTsx::add_second_p_a_i_hdr(pjsip_msg* msg)
...
+        {
+          std::string tel_URI = "tel:";
```
Comment (rkd-msw) [inline src/scscfsproutlet.cpp]: Is it clearer to call this `tel_prefix` or `tel_uri_prefix`?

File: src/scscfsproutlet.cpp [hunk 3]
```diff
@@ -1825,6 +1828,23 @@ void SCSCFSproutletTsx::add_second_p_a_i_hdr(pjsip_msg* msg)
...
+          std::size_t found = _aliases[i].rfind(tel_URI.c_str(), 4);
```
Comment (rkd-msw) [inline src/scscfsproutlet.cpp]: I would probably write this as `bool has_tel_prefix = (_aliases[i].rfind(tel_prefix, 4) != std::string::npos)` - your if test then becomes `if (has_tel_prefix)` which makes the overall intent of the code clearer.

File: src/scscfsproutlet.cpp [hunk 4]
```diff
@@ -1825,6 +1828,23 @@ void SCSCFSproutletTsx::add_second_p_a_i_hdr(pjsip_msg* msg)
...
       }
+      else
+      {
+        for (unsigned int i=0; i<_aliases.size(); i++)
+        {
+          std::string tel_URI = "tel:";
+          std::size_t found = _aliases[i].rfind(tel_URI.c_str(), 4); 
+          if (found!=std::string::npos)
+          {
+            TRC_DEBUG("Add second P-Assered Identity for %s", _aliases[i].c_str());
```
Comment (rkd-msw) [inline src/scscfsproutlet.cpp]: Typo - "Assered" -> "Asserted"

File: src/ut/scscf_test.cpp [hunk 5]
```diff
@@ -7056,6 +7056,25 @@ TEST_F(SCSCFTest, TestAddSecondTelPAIHdr)
   doSuccessfulFlow(msg, testing::MatchesRegex(".*wuntootreefower.*"), hdrs, false);
 }
 
+TEST_F(SCSCFTest, TestAddSecondTelPAIHdriWithAlias)
```
Comment (rkd-msw) [inline src/ut/scscf_test.cpp]: I think there's a typo in the test name - an extra 'i' in between 'HdrWith'.

Comment (rkd-msw) [inline src/ut/scscf_test.cpp]: It's generally good practice to add a short comment before a testcase, to make it clear what you're testing - in this case, that a tel: URI alias is added even when the username is different from the original SIP URI.

File: src/ut/scscf_test.cpp [hunk 6]
```diff
@@ -7075,6 +7094,24 @@ TEST_F(SCSCFTest, TestAddSecondSIPPAIHdr)
   doSuccessfulFlow(msg, testing::MatchesRegex(".*wuntootreefower.*"), hdrs, false);
 }
 
+TEST_F(SCSCFTest, TestAddSecondSIPPAIHdrNoSIPUri)
```
Comment (rkd-msw) [inline src/ut/scscf_test.cpp]: Same comment about having a comment before the test explaining its purpose - a sentence or two of normal writing is a lot better for that than a test name like `TestAddSecondSIPPAIHdrNoSIPUri`,

**Comment by rkd-msw** [comment]
A couple of comments inline. One other thing is that I think it's worth having more testcases to cover the more complicated cases, in particular:

* when you have multiple TEL URI aliases, none of which share the same userpart as the SIP URI, the first one in the list should be used
* when you have multiple TEL URI aliases, one of which shares the same userpart as the SIP URI, it should be used even if it's not the first one in the list

**Comment by AME2** [comment]
Applied mark-ups

**Comment by rkd-msw** [comment]
Couple of minor comments, good to merge once those are fixed.

**Pull request merged**